src/Discord.Net.WebSocket/DiscordSocketClient.cs

```diff
@@ -1410,6 +1420,13 @@ private async Task ProcessMessageAsync(GatewayOpCode opCode, int? seq, string ty
                                             guild.CompleteDownloadUsers();
                                             await TimedInvokeAsync(_guildMembersDownloadedEvent, nameof(GuildMembersDownloaded), guild).ConfigureAwait(false);
                                         }
+
+                                        if (data.Nonce.IsSpecified
+                                            && data.ChunkIndex + 1 >= data.ChunkCount
+                                            && _guildMembersRequestTasks.TryRemove(data.Nonce.Value, out var tcs))
```

Comment:
This being the only way of removing request tasks may result in a memory leak if client misses the event (due to temporary connection loss, discord requesting a reconnect, etc.)
I'm not exactly sure what's the best way around this, but I'd recommend adding a timeout to tasks so it retries/fails the task if the client doesn't receive the member chuck in X time. This would also solve the issue of possible deadlock caused by this

Reply:
What do you think about allowing users to pass a CancellationToken in `DownloadUsersAsync(IEnumerable<ulong> userIds)`? A fixed timeout seems a bit too inflexible IMHO, since downloading >1k users will actually take a while (=> more than a few seconds). If no CancellationToken is given, we could apply a default timeout

Reply:
Added a default request timeout and a way for users to supply their own cancellation token.